### PR TITLE
Git Harvest Plugin and Tutorial

### LINKED ISSUE #255: Git Plugin as Harvest Source
Relates to #246 
Write git plugin as harvest source

### PR DESCRIPTION
**Pull request by SKernchen**
Closes #255 .
This Feature shows a Tutorial for writing plugins for HERMES.  To do that the git harvest plugin will be in another project.
Currently in https://github.com/hermes-hmc/hermes-git

### CODE REVIEW

**Comment by SKernchen** [comment]
My idea is to open a new public repository named plugins/ and add harvest-git there. Then I would like to use this plugin in the workflow source code, so HERMES would actively use it.

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+<!--
+SPDX-FileContributor: Michael Meinel
+SPDX-FileContributor: Sophie Kernchen
+-->
+
+# Write a plugin for HERMES
+ 
+
+This tutorial will present the basic steps for writing an additional harvester.
+At the moment only the harvest architecture is stable.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:16]:
```suggestion
At the moment only the architecture for harvester plugins is stable.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+For this tutorial you should be familiar with HERMES. 
+If you never used HERMES before, you might want to check the tutorial: [Automated Publication with HERMES](https://docs.software-metadata.pub/en/latest/tutorials/automated-publication-with-ci.html).
+```
+
+## Plugin Architecture
+
+HERMES uses a plugin architecture. Therefore, users are invited to contribute own features.
+The structure for every plugin follows the same schema.
+There is a base class for every plugin. In this HermesPlugin class there is one abstract method __ call __ which needs to be overwritten.
+Furthermore, the HermesCommand class provides all needs for writing a plugin used in a HERMES command.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:30]:
```suggestion
Furthermore, the `HermesCommand` class provides all needs for writing a plugin used in a HERMES command.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+So the HermesPlugins call method uses an Instance of the HermesCommand that triggered this plugin to run.
+In our case this will be the HermesHarvestCommand which calls all harvest plugins.
+The Plugin class also uses a derivative of HermesSettings to add parameters.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:33]:
```suggestion
The plugin class also uses a derivative of `HermesSettings` to add parameters that can be adapted by the configuration file.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+HermesSettings are the base class for command specific settings.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:34]:
```suggestion
`HermesSettings` are the base class for command specific settings.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+It uses pydantic settings to specify and validate the parameters.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:35]:
```suggestion
It uses [pydantic](https://docs.pydantic.dev/latest/) [settings](https://docs.pydantic.dev/latest/api/pydantic_settings/) to specify and validate the parameters.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+The user can either set the parameters in the hermes.toml or overwrite them in the command line.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:36]:
```suggestion
The user can either set the parameters in the `hermes.toml` or overwrite them in the command line.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+To overwrite the configuration, you use the -O operator with the dotted parameter name and the value.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:37]:
```suggestion
To overwrite a parameter form command line, use the `-O` command line option followed the dotted parameter name and the value.
E.g., you can set your authentication token for InvenioRDM by adding the following options to your call to `hermes deposit`:

```shell
hermes deposit -O invenio_rdm.auth_token YourSecretAuthToken
```

Note that this is currently limited to plain parameters, i.e., passing strings and numbers but not lists or mappings.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+class GitHarvestPlugin(HermesHarvestPlugin):
+    settings_class = GitHarvestSettings
+
+    def __call__(self, command):
+        print("Hello World!")
+
+        return {}, {}
+```
+ 
+The Code uses the HermesHarvestPlugin as base class and pydantics Basemodel for the Settings. In the GitHarvestSettings you
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:64]:
```suggestion
The code uses the `HermesHarvestPlugin` as base class and pydantics Basemodel for the settings. In the `GitHarvestSettings` you
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+can see that one setting is made. The Parameter from_branch is specific for this plugin and can be reached through self.settings.harvest.git.git_branch as long as our plugin will be named git.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:65]:
```suggestion
can see that additional parameter is defined. The Parameter `from_branch` is specific for this plugin and can be accessed inside the plugin using `self.settings.harvest.git.git_branch` as long as our plugin will be named git.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+In the hermes.toml this would be achieved by [harvest.{plugin_name}].
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:66]:
```suggestion
In the `hermes.toml` this would be achieved by [harvest.{plugin_name}].
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+The GitHarvestSettings are assigned to the GitHarvestPlugin. In the plugin you need to overwrite the __ call __ method.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:67]:
```suggestion
The `GitHarvestSettings` are associated with the `GitHarvestPlugin`. In the plugin you need to overwrite the `__ call __` method.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+For now a simple Hello World will do. The method return two dictionaries. These will later depict the harvested data in codemeta (json-ld) and information for generating hermes metadata.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:68]:
```suggestion
For now a simple Hello World will do. The method returns two dictionaries. These will contain the harvested data in CodeMeta (JSON-LD) and information additional information, e.g., to provide provenance information.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+That is the basic structure for the plugins source code.
+
+To integrate this code, you have to register it as a plugin in the pyproject.toml. To learn more about the pyproject.toml check https://python-poetry.org/docs/pyproject/.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:71]:
```suggestion
To integrate this code, you have to register it as a plugin in the `pyproject.toml`. To learn more about the `pyproject.toml` check https://python-poetry.org/docs/pyproject/ or refer to [PEP621](https://peps.python.org/pep-0621/).
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+We will just look at the important places for this plugin. There are two ways to integrate this plugin. First we will show how to use the plugin environment as the running base with HERMES as a dependency.
+Then we say how to integrate this plugin in HERMES itself.
+
+### Include HERMES as Dependency
+This is probably the more common way, where you can see HERMES as a framework.
+The idea is that your project is the main part. You create the pyproject.toml as usual.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:77]:
```suggestion
The idea is that your project is the main part. You create the `pyproject.toml` as usual.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+In the dependencies block you need to include hermes. Then you just have to declare your plugin.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:78]:
```suggestion
In the dependencies block you need to include `hermes`. Then you just have to declare your plugin.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+The HERMES software will look for plugins and install these.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:79]:
```suggestion
The HERMES software will look for installed plugins and use them.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+In the code below you can see the parts of the pyproject.toml that are important.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:80]:
```suggestion
In the code below you can see the parts of the `pyproject.toml` that are important.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+[tool.poetry.dependencies]
+python = "^3.10"
+hermes = "^0.8.0"
+...
+...
+[tool.poetry.plugins."hermes.harvest"]
+git = "hermes_git.harvest:GitHarvestPlugin"
+...
+```
+As you can see the plugin class from hermes_git is declared as git for hermes.harvest.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:92]:
```suggestion
As you can see the plugin class from `hermes_git` is declared as `git` for the `hermes.harvest` entrypoint.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+To use the plugin you have to adapt the harvest Settings in the hermes.toml.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:93]:
```suggestion
To use the plugin you have to adapt the harvest settings in the `hermes.toml`.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+We will discuss the exact step after showing the other pyproject.toml configuration.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:94]:
```suggestion
We will discuss the exact step after showing the other `pyproject.toml` configuration.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+In the dependencies you have to install your plugin. If your Plugin is pip installable than you can just give the name and the version.
+If your plugin is in a buildable git repository, you can install it with the given expression. 
+Note that this differs with the accessibility and your wishes, check [Explicit Package Sources](https://python-poetry.org/docs/repositories/#explicit-package-sources).
+
+The second thing to adapt is to declare the access point for the plugin.
+You can do that with `git = "hermes_git.harvest:GitHarvestPlugin"`.
+This expression makes the GitHarvestPlugin from the hermes_git package, a hermes.harvest plugin named git.
+So you need to configure this line with your plugin properties.
+
+Now you just need to add the plugin to the hermes.toml and reinstall the adapted poetry package.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:127]: I think this is out of scope here.

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+
+### Configure hermes.toml
+To use the plugin, you have to set it in the hermes.toml.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:130]:
```suggestion
To use the plugin, you have to activate it in the `hermes.toml`.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+The settings for the plugins are also set there.
+For the harvest plugin the hermes.toml could look like this:
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:132]:
```suggestion
For the harvest plugin the `hermes.toml` could look like this:
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+sources = [ "cff", "git" ] # ordered priority (first one is most important)
+
+[harvest.cff]
+enable_validation = false
+
+[harvest.git]
+from_branch = "develop"
+...
+```
+In [harvest] you define that this plugin is used with less priority than the built-in cff plugin.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:144]:
```suggestion
In the `[harvest]` section you define that this plugin is used with less priority than the built-in `cff` plugin.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 25]
```diff
@@ -0,0 +1,156 @@
...
+in [harvest.git] you set the configuration for the plugin.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:145]:
```suggestion
in the `[harvest.git]` section you set the configuration for the plugin. 
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+[harvest.cff]
+enable_validation = false
+
+[harvest.git]
+from_branch = "develop"
+...
+```
+In [harvest] you define that this plugin is used with less priority than the built-in cff plugin.
+in [harvest.git] you set the configuration for the plugin. 
+In the Beginning of this tutorial we set the parameter `from_branch` in the git settings. Now we change the default from_branch to develop.
```
Comment (led02) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:146]:
```suggestion
In the beginning of this tutorial we set the parameter `from_branch` in the git settings. Now we change the default `from_branch` to `develop`.
```

**Review by led02 — CHANGES_REQUESTED** [review]
Mostly formatting, but some minor (content) changes are also proposed.
There is one line that tells the wrong story...

File: docs/source/tutorials/writing-a-plugin-for-hermes.md (hunk 7 above)
Comment (SKernchen) [reply docs/source/tutorials/writing-a-plugin-for-hermes.md:37]:
```suggestion
To overwrite a parameter form command line, use the `-O` command line option followed by the dotted parameter name and the value.
E.g., you can set your authentication token for InvenioRDM by adding the following options to your call to `hermes deposit`:
```shell
hermes deposit -O invenio_rdm.auth_token YourSecretAuthToken
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md (hunk 9 above)
Comment (SKernchen) [reply docs/source/tutorials/writing-a-plugin-for-hermes.md:65]:
```suggestion
can see that an additional parameter is defined. The Parameter `from_branch` is specific for this plugin and can be accessed inside the plugin using `self.settings.harvest.git.git_branch` as long as our plugin will be named git.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md (hunk 12 above)
Comment (SKernchen) [reply docs/source/tutorials/writing-a-plugin-for-hermes.md:68]:
```suggestion
For now a simple Hello World will do. The method returns two dictionaries. These will contain the harvested data in CodeMeta (JSON-LD) and additional information, e.g., to provide provenance information.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 27]
```diff
@@ -0,0 +1,156 @@
...
+The Code uses the HermesHarvestPlugin as base class and pydantics Basemodel for the Settings. In the GitHarvestSettings you
+can see that one setting is made. The Parameter from_branch is specific for this plugin and can be reached through self.settings.harvest.git.git_branch as long as our plugin will be named git.
+In the hermes.toml this would be achieved by [harvest.{plugin_name}].
+The GitHarvestSettings are assigned to the GitHarvestPlugin. In the plugin you need to overwrite the __ call __ method.
+For now a simple Hello World will do. The method return two dictionaries. These will later depict the harvested data in codemeta (json-ld) and information for generating hermes metadata.
+That is the basic structure for the plugins source code.
+
+To integrate this code, you have to register it as a plugin in the pyproject.toml. To learn more about the pyproject.toml check https://python-poetry.org/docs/pyproject/.
+We will just look at the important places for this plugin. There are two ways to integrate this plugin. First we will show how to use the plugin environment as the running base with HERMES as a dependency.
+Then we say how to integrate this plugin in HERMES itself.
```
[the remark this answers is not in the document]
Comment (SKernchen) [reply docs/source/tutorials/writing-a-plugin-for-hermes.md:73]: The idea was to show both ways, so that in the future contributors could add a pull request if they think, they have a really good plugin or a local hermes instance that they want to enhance...
I also like both for a better understanding on how the plugin structure works but only the first way would work too

File: docs/source/tutorials/writing-a-plugin-for-hermes.md (hunk 18 above)
Comment (SKernchen) [reply docs/source/tutorials/writing-a-plugin-for-hermes.md:92]: Maybe we should change that to hermes_plugin_git?

File: docs/source/tutorials/writing-a-plugin-for-hermes.md (hunk 27 above)
[the remark this answers is not in the document]
Comment (led02) [reply docs/source/tutorials/writing-a-plugin-for-hermes.md:76]: Okay, I'm fine with that. :)

**Comment by SKernchen** [comment]
We should maybe keep the name convention hermes-plugin-git through the whole project and not change to hermes_git but to hermes_plugin_git

**Review by led02 — APPROVED** [review]

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 28]
```diff
@@ -0,0 +1,158 @@
...
+In the code below you can see the parts of the `pyproject.toml` that are important.
+```{code-block} toml
+...
+[tool.poetry.dependencies]
+python = "^3.10"
+hermes = "^0.8.0"
+...
+...
+[tool.poetry.plugins."hermes.harvest"]
+git = "hermes_git.harvest:GitHarvestPlugin"
```
Comment (SKernchen) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:92]:
```suggestion
git = "hermes_plugin_git.harvest:GitHarvestPlugin"
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 29]
```diff
@@ -0,0 +1,158 @@
...
+...
+```
+As you can see the plugin class from `hermes_git` is declared as `git` for the `hermes.harvest` entrypoint.
```
Comment (SKernchen) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:95]:
```suggestion
As you can see the plugin class from `hermes_plugin_git` is declared as `git` for the `hermes.harvest` entrypoint.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 30]
```diff
@@ -0,0 +1,158 @@
...
+This variant is used to contribute to the HERMES community or adapt the HERMES workflow for own purposes.
+If you want to contribute, see the [Contribution Guidelines](https://docs.software-metadata.pub/en/latest/dev/contribute.html).
+After cloning the HERMES workflow repository you can adapt the pyproject.toml.
+In the code below you see the parts with the important lines.
+```{code-block} toml
+...
+[tool.poetry.dependencies]
+...
+pydantic-settings = "^2.1.0"
+hermes-git = { git = "https://github.com/hermes-hmc/hermes-git.git", branch = "main" }
```
Comment (SKernchen) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:112]:
```suggestion
hermes-plugin-git = { git = "https://github.com/softwarepub/hermes-plugin-git.git", branch = "main" }
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 31]
```diff
@@ -0,0 +1,158 @@
...
+...
+...
+[tool.poetry.plugins."hermes.harvest"]
+cff = "hermes.commands.harvest.cff:CffHarvestPlugin"
+codemeta = "hermes.commands.harvest.codemeta:CodeMetaHarvestPlugin"
+git = "hermes_git.harvest:GitHarvestPlugin"
```
Comment (SKernchen) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:118]:
```suggestion
git = "hermes_plugin_git.harvest:GitHarvestPlugin"
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 32]
```diff
@@ -0,0 +1,158 @@
...
+...
+```
+In the dependencies you have to install your plugin. If your Plugin is pip installable than you can just give the name and the version.
+If your plugin is in a buildable git repository, you can install it with the given expression. 
+Note that this differs with the accessibility and your wishes, check [Explicit Package Sources](https://python-poetry.org/docs/repositories/#explicit-package-sources).
+
+The second thing to adapt is to declare the access point for the plugin.
+You can do that with `git = "hermes_git.harvest:GitHarvestPlugin"`.
```
Comment (SKernchen) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:126]:
```suggestion
You can do that with `git = "hermes_plugin_git.harvest:GitHarvestPlugin"`.
```

File: docs/source/tutorials/writing-a-plugin-for-hermes.md [hunk 33]
```diff
@@ -0,0 +1,158 @@
...
+This expression makes the GitHarvestPlugin from the hermes_git package, a hermes.harvest plugin named git.
```
Comment (SKernchen) [inline docs/source/tutorials/writing-a-plugin-for-hermes.md:127]:
```suggestion
This expression makes the GitHarvestPlugin from the hermes_plugin_git package, a hermes.harvest plugin named git.
```
